### DISCUSSION TITLE
How to create a circular hole in square by mesh generator

### DISCUSSION THREAD

**Discussion by Clark-domain** [idaholab/moose#18557] [Q&A General] [answered]
Hi, team,

I wanted to create a circular hole in the square by mesh generator. I  used the following blocks from the input file, but the result was not good as seen below in Fig. 1. The expected result should be like a hole in Fig. 2.  Could you give me some advice on how to make it? Thanks very much.

 [./gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 30
    ny = 30
    xmax = 60
    ymax = 60
    elem_type = TRI3
  [../]
  [./hole]
    input = gen
    type = ParsedSubdomainMeshGenerator
    combinatorial_geometry = 'r:=sqrt((x-30)^2+(y-30)^2);if(r<=15,1,0)'
    block_id = 1
  [../]   
  [block_deleted]
    type = BlockDeletionGenerator
    input = hole
    block = 1
  []

Fig.1
![image](https://user-images.githubusercontent.com/74393696/128660473-81511ce1-74f3-472c-8fc5-cb9d14913967.png)
Fig.2
![image](https://user-images.githubusercontent.com/74393696/128660187-4769c0ec-dc3f-4cdd-b3cd-17b4c09ec2d9.png)

**Comment by GiudGiud** [comment, marked as answer]
Hello

Use this mesh generator with the outer_box parameter
https://mooseframework.inl.gov/source/meshgenerators/ConcentricCircleMeshGenerator.html
then delete the center circle. 

Guillaume

**Reply by Clark-domain** [reply]
Hi,

I wanted to create a hole on a rectangular plate instead of a square plate. I found it is not straightforward. Can you give me some suggestions?

Thanks very much,

Kaijie

**Reply by GiudGiud** [reply]
You can stitch meshes on the sides on the square mesh to make it a rectangle.

**Reply by Clark-domain** [reply]
I am trying to solve it from your suggestion. I stitched the other one on the right side. [See below]. But now there is a boundary problem.

I wanted to apply pressure to the right, but actually, it was on the top. The input and mesh info are shown below. Can you help me? Thanks very much.

[Mesh]
  [./square] 
    type = ConcentricCircleMeshGenerator
    num_sectors = 10
    radii = '    10    20'                
    rings = '4     10     6'            
    has_outer_square = on
    pitch = 60
    preserve_volumes = off
    smoothing_max_it = 8
  []
  
  [square_hole]  
    type = BlockDeletionGenerator
    input = square
    block = 1
  [] 

 [./rect_top]
  type = GeneratedMeshGenerator 
  dim = 2
  ny = 24
  nx = 10
  xmin = 30
  xmax = 50
  ymin = -30
  ymax = 30
 []
  
  [./smg]
    type = StitchedMeshGenerator
    inputs = 'square_hole rect_top '
    clear_stitched_boundary_ids = true
    stitch_boundaries_pairs = 'right left'
    parallel_type = 'replicated'
    show_info = true
  []  
[]

In ReplicatedMesh::stitch_meshes:
This mesh has 27 nodes on boundary 1.
Other mesh has 27 nodes on boundary 3.
Minimum edge length on both surfaces is 1.
In ReplicatedMesh::stitch_meshes:
Found 27 matching nodes.

In ReplicatedMesh::stitch_meshes:
This mesh has 53 nodes on boundary 2.
Other mesh has 53 nodes on boundary 4.
Minimum edge length on both surfaces is 1.
In ReplicatedMesh::stitch_meshes:
Found 53 matching nodes.

In ReplicatedMesh::stitch_meshes:
This mesh has 25 nodes on boundary 3.
Other mesh has 25 nodes on boundary 3.
Minimum edge length on both surfaces is 2.5.
In ReplicatedMesh::stitch_meshes:
Found 25 matching nodes.

StitchedMeshGenerator 'smg':  Mesh Information:
StitchedMeshGenerator 'smg':   elem_dimensions()={2}
StitchedMeshGenerator 'smg':   spatial_dimension()=2
StitchedMeshGenerator 'smg':   n_nodes()=1194
StitchedMeshGenerator 'smg':     n_local_nodes()=325
StitchedMeshGenerator 'smg':   n_elem()=1116
StitchedMeshGenerator 'smg':     n_local_elem()=279
StitchedMeshGenerator 'smg':     n_active_elem()=1116
StitchedMeshGenerator 'smg':   n_subdomains()=3
StitchedMeshGenerator 'smg':   n_partitions()=4
StitchedMeshGenerator 'smg':   n_processors()=4
StitchedMeshGenerator 'smg':   n_threads()=1
StitchedMeshGenerator 'smg':   processor_id()=0
StitchedMeshGenerator 'smg':   is_prepared()=true
StitchedMeshGenerator 'smg':   is_replicated()=true
StitchedMeshGenerator 'smg': 
StitchedMeshGenerator 'smg':  Mesh Bounding Box:
StitchedMeshGenerator 'smg':   Minimum: (x,y,z)=(     -30,      -30,        0)
StitchedMeshGenerator 'smg':   Maximum: (x,y,z)=(      50,       30,        0)
StitchedMeshGenerator 'smg':   Delta:   (x,y,z)=(      80,       60,        0)
StitchedMeshGenerator 'smg': 
StitchedMeshGenerator 'smg':  Mesh Element Type(s):
StitchedMeshGenerator 'smg':   QUAD4
StitchedMeshGenerator 'smg': 
StitchedMeshGenerator 'smg':  Mesh Nodesets:
StitchedMeshGenerator 'smg':   Nodeset 0 (bottom), 10 nodes
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(      32,      -30,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
StitchedMeshGenerator 'smg':   Nodeset 1 (left), 25 nodes
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(      50,      -30,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
StitchedMeshGenerator 'smg':   Nodeset 2 (bottom), 10 nodes
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(      32,       30,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
StitchedMeshGenerator 'smg': 
StitchedMeshGenerator 'smg':  Mesh Sidesets:
StitchedMeshGenerator 'smg':   Sideset 0 (bottom), 10 sides (EDGE2), 10 elems (QUAD4), 11 nodes
StitchedMeshGenerator 'smg':    Side volume: 20
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(      30,      -30,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(      50,      -30,        0)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(      20, 7.10543e-15,        0)
StitchedMeshGenerator 'smg':   Sideset 1 (left), 48 sides (EDGE2), 48 elems (QUAD4), 50 nodes
StitchedMeshGenerator 'smg':    Side volume: 120
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(     -30,      -30,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(      50,       30,        0)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(      80,       60,        0)
StitchedMeshGenerator 'smg':   Sideset 2 (bottom), 34 sides (EDGE2), 34 elems (QUAD4), 36 nodes
StitchedMeshGenerator 'smg':    Side volume: 80
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(     -30,      -30,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(      50,       30,        0)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(      80,       60,        0)
StitchedMeshGenerator 'smg':   Sideset 4 (top), 24 sides (EDGE2), 24 elems (QUAD4), 25 nodes
StitchedMeshGenerator 'smg':    Side volume: 60
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(     -30,       30,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(      30,       30,        0)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(      60, 3.55271e-15,        0)
StitchedMeshGenerator 'smg': 
StitchedMeshGenerator 'smg':  Mesh Edgesets:
StitchedMeshGenerator 'smg':   None
StitchedMeshGenerator 'smg': 
StitchedMeshGenerator 'smg':  Mesh Subdomains:
StitchedMeshGenerator 'smg':   Subdomain 0: 240 elems (QUAD4, 240 active), 275 active nodes
StitchedMeshGenerator 'smg':    Volume: 1200
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(      30,      -30,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(      50,       30,        0)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(      20,       60,        0)
StitchedMeshGenerator 'smg':   Subdomain 2: 400 elems (QUAD4, 400 active), 437 active nodes
StitchedMeshGenerator 'smg':    Volume: 938.607
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(     -20,      -20,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(      20,       20,        0)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(      40,       40,        0)
StitchedMeshGenerator 'smg':   Subdomain 3: 476 elems (QUAD4, 476 active), 541 active nodes
StitchedMeshGenerator 'smg':    Volume: 2348.52
StitchedMeshGenerator 'smg':    Bounding box minimum: (x,y,z)=(     -30,      -30,        0)
StitchedMeshGenerator 'smg':    Bounding box maximum: (x,y,z)=(      30,       30,        0)
StitchedMeshGenerator 'smg':    Bounding box delta: (x,y,z)=(      60,       60,        0)
StitchedMeshGenerator 'smg':   Global mesh volume = 4487.13
The following total 1 aux variables:
  Fe
are added for automatic output by MaterialOutputAction.




![image](https://user-images.githubusercontent.com/74393696/131219621-41e6576b-8d1e-4d8b-a42b-377cfee57d5b.png)

**Reply by GiudGiud** [reply]
Those mesh generators have pre-defined names and maybe the stitched ones dont match so well. 
I would re-define the boundary you want using the sidesetmesh generators

**Reply by Clark-domain** [reply]
Here is what I changed in the input file. But it showed an error. How should I further improve it? Thanks.

![image](https://user-images.githubusercontent.com/74393696/131221871-4f118c78-a3da-4f26-b656-ab13c447a11c.png)


[Mesh]
 [./rectR]
  type = GeneratedMeshGenerator 
  dim = 2
  ny = 24
  nx = 10
  xmin = 30
  xmax = 50
  ymin = -30
  ymax = 30
 []
 [./square]  
    type = ConcentricCircleMeshGenerator
    num_sectors = 10
    radii = '    10    20'                  
    rings = '4     10     6'              
    has_outer_square = on
    pitch = 60
    preserve_volumes = off
    smoothing_max_it = 8
  []  
  [squareHole]  
    type = BlockDeletionGenerator
    input = square
    block = 1
  [] 

 [./smg]
    type = StitchedMeshGenerator
    inputs = 'squareHole rectR '
    clear_stitched_boundary_ids = true
    stitch_boundaries_pairs = 'right left'
    parallel_type = 'replicated'
    show_info = true
  []
  [./block_1]
    type = SideSetsAroundSubdomainGenerator
    input = smg
    block = 'rectR'
    new_boundary = 'pull'
    normal = '1 0 0'
  []
  [./block_2]
    type = SideSetsAroundSubdomainGenerator
    input = block_1
    block = 'squareHole'
    new_boundary = 'fix'
    normal = '-1 0 0'
  [] 

  construct_side_list_from_node_list = true  
[]

**Reply by GiudGiud** [reply]
Can you run with `--mesh-only` mode, look at the mesh in paraview then examine what sidesets seem to be missing?
What is the name of the sideset you are trying to reference in the boundary condition?

**Reply by Clark-domain** [reply]
Great. I am doing it. It is getting better. 
With ParaView, I can see the element block ID (fig 1 below), but I cannot see boundary ID(fig 2 below). Where should I see boundary ID? Thanks.

![image](https://user-images.githubusercontent.com/74393696/131231711-5bf77bb0-5368-4c75-ae37-4e529794e804.png)
![image](https://user-images.githubusercontent.com/74393696/131231725-b55f8a34-2bd1-4263-9285-78d87ab9a5f1.png)

**Reply by GiudGiud** [reply]
which sideset is the boundary condition meant to be on ?
Any sideset missing in what you see in Paraview?

You should be able to see sidesets under `Sets` in the `Properties` tab

**Reply by Clark-domain** [reply]
Now I can see the boundary from your suggestion. 

I  noticed that, after stitching the meshes, though no boundary ids are missing, I could not apply the boundary id 3 (left). The error occurred:
![image](https://user-images.githubusercontent.com/74393696/131279689-d3bd64af-1ad8-4e34-b2fb-3d36ca67927c.png)
Then I further added AllSideSetsByNormalsGenerator, and then I can apply the boundary id 3. Why is that? When should I use AllSideSetsByNormalsGenerator ? Thanks very much.

**Reply by GiudGiud** [reply]
Those are probably removed because they are likely to not be valid anymore. Imagine if you had stitched on the left rather on the right, `left` would be in the middle of the domain. 
Some generators will re-do those boundaries after they are applied I think

**Reply by Clark-domain** [reply]
Sure. I appreciate your help. I will pay attention to it in the future. Now I have worked it out. I tested the max stress near the hole on the plate and it gave the same value as the analytical result. Thanks very much.

**Comment by Clark-domain** [comment]
Great. It worked. I just made a 60x60 square with a 20 diameter hole. See below. Thanks very much.


![image](https://user-images.githubusercontent.com/74393696/128803655-9885a180-cfcf-482e-8987-f8772338e081.png)
